### DISCUSSION TITLE
What is the best way to implement multi-language in XAML app?

### DISCUSSION THREAD

**Discussion by panthernet** [AvaloniaUI/Avalonia#16445] [Q&A] [answered]
I want to be able to change language on the fly.  I have two variants in my mind:
1. Bind Text to static dictionary on App class which holds all strings. Not really follows the DI principes.
Something like this: ` <TextBlock Text="{Binding Source={x:Static app:App.Language}, Path=Translations[TTT], FallbackValue='Server address and port'}" />`

2. Use Locator to get the same dictionary as service into VM and bind form XAML to it. Can this lead to binding leaks?

Maybe somethign else? What is the best case?

**Comment by panthernet** [comment]
I guess i should be using smth like this https://www.sakya.it/wordpress/avalonia-ui-framework-localization/
To get the refresh on the fly.

**Comment by maxkatz6** [comment]
To begin, there is no best way. But many opinions and options.

As my opinion, I don't think language should rely on DI, as any other View-layer part of the codebase. Your first option is perfectly fine, although I would hide it under custom markup extension that returns the same Binding from ProvideValue, with single key parameter.

If you insist, the same markup extension could use DI locator, yes.

> Can this lead to binding leaks?

Bindings themselves shouldn't leak in any case. But control instances might, but you just shouldn't keep static strong references on them in the code.

**Comment by kyuranger** [comment, marked as answer]
Firsly, I created an extension like this:
```
  public class LocalizationExtension : MarkupExtension
  {
      public override object ProvideValue(IServiceProvider serviceProvider)
      {
             return LocalizationHelper.GetLocalizedString(_Key);
      }
      string _Key;
      public LocalizationExtension(string key)
      {
          _Key = key;
      }
  }
```

Then I created a helper:
```
  public class LocalizationHelper 
  {
      static FrozenDictionary<string, string> s_localizationDictionary = FrozenDictionary<string, string>.Empty;
      public static string GetLocalizedString(string key)
      {
          if (s_localizationDictionary is { } && s_localizationDictionary.TryGetValue(key, out var _result))
          {
              return _result;
          }
          else
          {
              return key;
          }
      }      
      public void Initialize(string languageName)
      {
         ///Initialize s_localizationDictionary
      }       
  }
```

I stored multi-language data in different json files, such as en.json. you must use the Initialize method in LocalizationHelper to initialize multi-language data. The key in the dictionary is the only key, and the value is the translation of the corresponding language. For example, if the key is HelloWorld, the corresponding value for English is Hello World, and the corresponding value for Chinese is 你好世界.

ps: FrozenDictionary is a new feature of .NET8. It is more efficient to use FrozenDictionary for fixed data. Of course, if you are not using .NET8, you can use the normal Dictionary.

In XAML, you can use it like this:
`<TextBlock Text="{Extensions:Localization HelloWorld}"></TextBlock>`
ps: Remember to define Extensions, for example, xmlns:Extensions="using:Sample.Extensions"

You might ask, why did I split it into two separate classes?

Because on the backend, if you want to get multilingual data, you can do it like this:
`LocalizationHelper.GetLocalizedString("HelloWorld");`

Although my method is feasible, it has a fatal flaw, because it does not use binding on the front end, so when the language is changed, the program needs to be restarted for it to take effect.
